### PR TITLE
opt.sh: refine option definition and overriding

### PR DESCRIPTION
**Pull request by aiChaoSONG**
1. move common option definition to a function, common options can be define here(eg, help, and tplg). TODO, call this in script explicitly(rather than in opt.sh), or some options may be hidden.
2. split  function in_func_create_tmpbash to _fill_opt_vars and _validate_and_format_cmd_line for better code structure.
3. fix unclear comments and add more comments

### CODE REVIEW

File: case-lib/opt.sh [hunk 1]
```diff
@@ -13,48 +13,46 @@ declare -A OPT_NAME OPT_HAS_ARG OPT_VAL OPT_DESC
...
+            }
         done
+    }
 
-        if  ! _op_temp_script=$(
-                getopt -o "$_short_opt" --long "$_long_opt" -- "$@"); then
-            # here output for the End-User who don't care about this function/option couldn't to run
-            printf 'Error parsing options: %s\n' "$*"
-            # To debug uncomment these lines:
-            # printf 'DEBUG parsing options: short_opt: %s / long_opt: %s / args: %s\n' "$_short_opt" "$_long_opt" "$*" >&2
```
Comment (marc-hb) [inline case-lib/opt.sh:62]: I left this `printf DEBUG` because it found it very useful, please keep it.

File: case-lib/opt.sh [hunk 2]
```diff
@@ -98,21 +96,26 @@ func_opt_parse_option()
...
+    # the option map used for OPT_VAL overriding
+    local -A short_opt_lst long_opt_lst
+
+    _fill_opt_vars
+    _validate_and_format_cmd_line "$@"
 
-    # FIXME: what is this supposed to do!?
-    eval set -- "$_op_temp_script"
+    # set the contents in formatted_cmd_opts to position arguments $1, $2, ...
+    eval set -- "$formatted_cmd_opts"
```
Comment (marc-hb) [inline case-lib/opt.sh:115]: Thanks for the new comment! Can you add a (small) `"$formatted_cmd_opts"` example here?

File: case-lib/opt.sh [hunk 3]
```diff
@@ -13,48 +13,46 @@ declare -A OPT_NAME OPT_HAS_ARG OPT_VAL OPT_DESC
...
-    # Asks getopt to validate command line input and generate $_op_temp_script
-    #
-    # Also initialize the global _op_short_list and _op_long_list
-    # "reversed maps" (reversed compared to the user's OPT_ maps) that
-    # map valid command line input to the corresponding
-    # one-character option code name. Example:
-    #   _op_short_lst=
-    #        ( [-t]="t" [-d]="d" [-l]="l" [-h]="h" )
-    #   _op_long_lst=
-    #        ( [--help]="h" [--duration]="d" [--tplg]="t" [--loop]="l" )
```
Comment (marc-hb) [inline case-lib/opt.sh:39]: I don't see your replacement for this comment as an improvement, you removed a fair bit for no clear reason and you broke the indentation.

**Review by marc-hb — CHANGES_REQUESTED** [review]
Thanks for cleaning up this code, I always found it very hard to follow.

I see there are only 2 shellcheck warnings left, they seem easy to fix and you're already touching that code anyway.


(pylint failed because of some apt-get failure; there's no python code change in this PR)

File: case-lib/opt.sh (hunk 3 above)
Comment (xiulipan) [reply case-lib/opt.sh:39]: I think the intend is fixed in `_func_case_dump_description`

File: case-lib/opt.sh [hunk 4]
```diff
@@ -71,71 +60,93 @@ func_opt_parse_option()
...
-            printf '\tthis message\n'
-            _func_case_dump_description
-            trap - EXIT
-            exit 2
-        }
+        printf '    -h |  --help\n'
+        printf '\tthis message\n'
+        _func_case_dump_description
+        trap - EXIT
+        exit 1
```
Comment (xiulipan) [inline case-lib/opt.sh:91]: I am thinking about we return as error with help message?

**Review by xiulipan — CHANGES_REQUESTED** [review]

File: case-lib/opt.sh (hunk 4 above)
Comment (aiChaoSONG) [reply case-lib/opt.sh]: @xiulipan agreed on this. this function is called if `-h` is specified, or command line parsing error. we should exit success for the first condition and failure for the second condition.

**Review by xiulipan — DISMISSED** [review]

**Comment by aiChaoSONG** [comment]
This failures on TGLH and BYT has nothing to do with this code change.

File: case-lib/opt.sh [hunk 5]
```diff
@@ -146,5 +163,15 @@ func_opt_parse_option()
...
+}
+
+# Define common options among test scripts.
+add_common_options()
+{
+    # The help option
+    OPT_NAME['h']='help'
+    OPT_HAS_ARG['h']=0
+    OPT_VAL['h']=0
+    OPT_DESC['h']='show help information'
```
Comment (marc-hb) [inline case-lib/opt.sh:176]: Can you move this function back to the top of the file? It's a great example so a great way to start the file.

File: case-lib/opt.sh [hunk 6]
```diff
@@ -1,141 +1,158 @@
...
+                [ "X${OPT_HAS_ARG[$i]}" != "X1" ] || printf ' parameter'
+            fi
+            printf '\n\t%s\n' "${OPT_DESC[$i]}"
+            if [ "${OPT_VAL[$i]}" ]; then
+                if [ "${OPT_HAS_ARG[$i]}" -eq 1 ]; then
+                    printf '\tDefault Value: %s\n' "${OPT_VAL[$i]}"
+                elif [ "${OPT_VAL[$i]}" -eq 0 ]; then
+                    printf '\tDefault Value: Off\n'
+                else
+                    printf '\tDefault Value: On\n'
```
Comment (marc-hb) [inline case-lib/opt.sh:84]: Next time please don't mix in the same PR such a big whitespace change with actual code changes because it's practically impossible to see whether any code has changed there.

**Review by marc-hb — APPROVED** [review]

**Comment by xiulipan** [comment]
@aiChaoSONG Please try to update the simple fix as @marc-hb required and merge or merge directly and send FIX later.

**Review by marc-hb — APPROVED** [review]

**Review by xiulipan — APPROVED** [review]
